### PR TITLE
Add CherryPy stats to WebTools/RESTModel based server

### PR DESCRIPTION
**Pull request by vkuznet**
The code provides ability to expose CherryPy statistics from https://cherrypydocrework.readthedocs.io/pkg/cherrypy.lib.html#module-cherrypy.lib.cpstats

Any web service which depends on WebTools/RESTModel class (e.g. DBS) will now serve CherryPy statistics via /stats endpoint, e.g. `cmsweb.cern.ch/dbs/stats` will provide this. The statistics dict will provide various metrics about underlying cherrypy activity. Here is an example of stats dict:
```
{
    "Cheroot HTTPServer 4517118096": {
        "Accepts": 0,
        "Accepts/sec": 0.0,
        "Bind Address": "('127.0.0.1', 8080)",
        "Bytes Read": -1,
        "Bytes Written": -1,
        "Enabled": false,
        "Queue": 0,
        "Read Throughput": -1,
        "Requests": -1,
        "Run time": -1,
        "Socket Errors": 0,
        "Threads": 10,
        "Threads Idle": 9,
        "Work Time": -1,
        "Worker Threads": {
            "CP Server Thread-10": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-11": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-12": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-3": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-4": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-5": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-6": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-7": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-8": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            },
            "CP Server Thread-9": {
                "Bytes Read": 0,
                "Bytes Written": 0,
                "Read Throughput": 0.0,
                "Requests": 0,
                "Work Time": 0,
                "Write Throughput": 0.0
            }
        },
        "Write Throughput": -1
    },
    "CherryPy Applications": {
        "Bytes Read/Request": 0.0,
        "Bytes Read/Second": 0.0,
        "Bytes Written/Request": 0.0,
        "Bytes Written/Second": 0.0,
        "Current Requests": 1,
        "Current Time": 1558095776.530752,
        "Enabled": true,
        "Requests": {
            "123145431896064": {
                "Bytes Read": null,
                "Bytes Written": null,
                "Client": "127.0.0.1:56683",
                "End Time": null,
                "Processing Time": 0.0001900196075439453,
                "Request-Line": "GET /stats HTTP/1.1",
                "Response Status": null,
                "Start Time": 1558095776.530613
            }
        },
        "Requests/Second": 0.4752343140987292,
        "Server Version": "17.4.1",
        "Start Time": 1558095774.426585,
        "Total Bytes Read": 0,
        "Total Bytes Written": 0,
        "Total Requests": 1,
        "Total Time": 0,
        "Uptime": 2.1041579246520996
    }
}
```

Once we'll have this in place this statistics can be captured by Prometheus exporters and we'll get them in Grafana dashboards. To do that I need to write new cherrypy exporter to get this metrics.

### CODE REVIEW

**Comment by vkuznet** [comment]
@amaltaro , @bbockelm please review this. I found a way to make it generic that it will be applied to any WebTools/RESTModel based servers, e.g. DBS will get it for free, CRAB (if it relies on this model).

**Comment by vkuznet** [comment]
Alan, I think it would be nice to include this in next cmsweb upgrade to enable cherrypy monitoring along with changes we'll put with WMCore and DBS. But I'll leave it up to you to decide about this PR.

**Comment by amaltaro** [comment]
Any chance you can figure out how the cherrypy application performance gets affected by this?
It's great to have all this monitoring, but at the same time we need to keep in mind that those don't come for free.

**Comment by vkuznet** [comment]
Alan,
my understanding that until someone ask for it nothing is executed, so the
penalty is zero. Of course, we'll ask for it through exporter on regular
interval, e.g. every 10 sec. But I think all the metrics are read from internal
objects rather then "extra" calculations/call functions. Since it is
configurable I can change this line:
cherrypy.config.update({'tools.cpstats.on': True})
to read from external configuration, e.g.
cherrypy.config.update({'tools.cpstats.on': configDict.get("cpstats", False)})
and then we can turned off if necessary via configuration parameter.

**Comment by vkuznet** [comment]
Alan, I made a change to make it optional and by default it is not enabled

**Comment by amaltaro** [comment]
Valentin, I can't really test these changes under WebTools because we don't use them in WMCore central services.

I made further local changes to also support it under the REST model, can you please update this PR with the following extra changes:
```
--- WMCore/REST/Server.py.old	2019-05-21 11:52:11.000000000 +0200
+++ WMCore/REST/Server.py.new	2019-05-21 14:44:28.205275476 +0200
@@ -10,7 +10,7 @@ from functools import wraps
 from threading import Thread, Condition
 
 from cherrypy import engine, expose, request, response, HTTPError, HTTPRedirect, tools
-
+from cherrypy.lib import cpstats
 from WMCore.REST.Error import *
 from WMCore.REST.Format import *
 from WMCore.REST.Validation import validate_no_more_input
@@ -354,6 +354,11 @@ class RESTFrontPage:
         URL arguments; they are not used here."""
         return self._serve([self._frontpage])
 
+    @expose
+    def stats(self):
+        "Return CherryPy stats dict about underlying service activities"
+        return cpstats.StatsPage().data()
+
 
 ######################################################################
 ######################################################################
@@ -690,6 +695,11 @@ class MiniRESTApi:
         self.methods[method][api] = apiobj
 
     @expose
+    def stats(self):
+        "Return CherryPy stats dict about underlying service activities"
+        return cpstats.StatsPage().data()
+
+    @expose
     def default(self, *args, **kwargs):
         """The HTTP request handler.
```

**Review by amaltaro — CHANGES_REQUESTED** [review]
And for the record, I updated the deployment configuration changes accordingly, see:
https://github.com/dmwm/deployment/pull/771

@yuyiguo any chance you can test this PR before we merge

**Comment by vkuznet** [comment]
Alan, please review/test the REST/Server changes for cpstats.

**Comment by amaltaro** [comment]
Thanks, Valentin. Changes look good, please squash these commits.

**Comment by vkuznet** [comment]
done.

**Comment by yuyiguo** [comment]
I will do phys03Writer or phys03Reader as the first try.

**Comment by amaltaro** [comment]
test this please

**Review by amaltaro — APPROVED** [review]

**Comment by amaltaro** [comment]
Tested on the WMCore land. DBS needs to make sure WebTools changes work as expected as well.

**Comment by yuyiguo** [comment]
@amaltaro 

This PR is not in 1.1.16_dbs. Could you merge it there?

**Comment by amaltaro** [comment]
@yuyiguo here is the backport: https://github.com/dmwm/WMCore/pull/9232
merged and in the tag 1.1.16.dbs5

Please review to make sure everything you need is in place.

**Pull request merged**
